bioimageio/spec/utils/build_spec.py

```diff
@@ -0,0 +1,246 @@
+    root: Optional[str] = None,
+    # model specific optional
+    weight_type: Optional[str] = None,
+    sample_inputs: Optional[str] = None,
+    sample_outputs: Optional[str] = None,
+    # tensor specific
+    input_name: Optional[str] = None,
+    input_step: Optional[List[int]] = None,
+    input_min_shape: Optional[List[int]] = None,
+    output_name: Optional[str] = None,
+    output_reference: Optional[str] = None,
+    output_scale: Optional[List[int]] = None,
+    output_offset: Optional[List[int]] = None,
+    halo: Optional[List[int]] = None,
+    preprocessing: Optional[List[Dict[str, Dict[str, Union[int, float, str]]]]] = None,
+    postprocessing: Optional[List[Dict[str, Dict[str, Union[int, float, str]]]]] = None,
+    # general optional
+    git_repo: Optional[str] = None,
+    attachments: Optional[List[str]] = None,
+    packaged_by: Optional[List[str]] = None,
+    run_mode: Optional[str] = None,
+    parent: Optional[str] = None,
+    config=None  # TODO not sure what are the correct type hints for config
+):
+    """
+    """
+    #
+    # generate the model specific fields
+    #
+
+    # check the test inputs and auto-generate input/output description from test inputs/outputs
+    for test_in, test_out in zip(test_inputs, test_outputs):
+        test_in, test_out = _ensure_uri(test_in, root), _ensure_uri(test_out, root)
+        test_in, test_out = np.load(test_in), np.load(test_out)
+    inputs = _get_input_tensor(test_in, input_name, input_step, input_min_shape, preprocessing)
+    outputs = _get_output_tensor(test_out, output_name,
+                                 output_reference, output_scale, output_offset,
+                                 postprocessing, halo)
+
+    (weights, language,
+     framework, source_hash) = _get_weights(weight_uri, weight_type,
+                                            source, root)
+
+    #
+    # generate general fields
+    #
+    format_version = '0.3.1'  # TODO get this from somewhere central
+    timestamp = datetime.datetime.now()
+
+    if source is not None:
+        source = fields.ImportableSource().deserialize(source)
+
+    # optional kwargs, don't pass them if none
+    optional_kwargs = {'git_repo': git_repo, 'attachments': attachments,
+                       'packaged_by': packaged_by, 'parent': parent,
+                       'run_mode': run_mode, 'config': config,
+                       'sample_inputs': sample_inputs,
+                       'sample_outputs': sample_outputs}
+    kwargs = {
+        k: v for k, v in optional_kwargs.items() if v is not None
+    }
+
+    # build the citation object
+    cite = _build_cite(cite)
+
+    model = raw_nodes.Model(
+        source=source,
+        sha256=source_hash,
+        kwargs=model_kwargs,
+        format_version=format_version,
+        name=name,
+        description=description,
+        authors=authors,
+        cite=cite,
+        tags=tags,
+        license=license,
+        documentation=documentation,
```

Comment:
should be `URI`, not `str`

Reply:
Seems to work implicitly.

Reply:
> Maybe the best solution would be to include a round-trip in this function.

Yes, that would make sense. What's the best function for this though?
Ideally we would need something that does only formal verification, i.e. are the URIs well formed etc. but does not check if they actually exist etc. Otherwise we limit how this function can be used by forcing the user that all the data is in place already.

Reply:
I thought of a round trip serialization/deserialization like so: https://github.com/bioimage-io/python-bioimage-io/pull/64/files/ce10c3bf1b59e80d2f07e4e876e3420c951b95e5#r639518930